shared/src/error/mod.rs

```diff
@@ -1,101 +1,116 @@
+use serde::{Deserialize, Serialize};
 use thiserror::Error;
 
-use crate::encoding::repr::*;
-
-/// ExitCode defines the exit code from the VM execution.
-#[repr(u32)]
-#[derive(
-    PartialEq, Eq, Debug, Clone, Copy, FromPrimitive, Serialize_repr, Deserialize_repr, Error,
-)]
-pub enum ExitCode {
-    Ok = 0,
-
-    /// Indicates failure to find an actor in the state tree.
-    SysErrSenderInvalid = 1,
+/// ExitCode defines the exit code from the VM invocation.
+#[derive(PartialEq, Eq, Debug, Clone, Copy, Serialize, Deserialize)]
+#[serde(transparent)]
+pub struct ExitCode {
+    value: u32,
+}
 
-    /// Indicates that the message sender was not in a valid state to send this message.
-    ///
-    /// Either:
-    /// - The sender's nonce nonce didn't match the message nonce.
-    /// - The sender didn't have the funds to cover the message gas.
-    SysErrSenderStateInvalid = 2,
+impl ExitCode {
+    /// The code indicating successful execution.
+    pub const OK: ExitCode = ExitCode::new(0);
 
-    /// Indicates failure to find a method in an actor.
-    SysErrInvalidMethod = 3,
+    /// The lowest exit code that an actor may abort with.
+    pub const FIRST_UNRESERVED_EXIT_CODE: u32 = 16;
 
-    /// Used for catching panics currently.
-    SysErrActorPanic = 4,
+    pub const fn new(value: u32) -> Self {
+        Self { value }
+    }
 
-    /// Indicates that the receiver of a message is not valid (and cannot be implicitly created).
-    SysErrInvalidReceiver = 5,
+    pub fn value(self) -> u32 {
+        self.value
+    }
 
-    /// Indicates a message sender has insufficient funds for a message's execution.
-    SysErrInsufficientFunds = 6,
+    /// Returns true if the exit code indicates success.
+    pub fn is_success(self) -> bool {
+        self.value == 0
+    }
 
-    /// Indicates message execution (including subcalls) used more gas than the specified limit.
-    SysErrOutOfGas = 7,
+    /// Returns true if the error code is in the range of exit codes reserved for the VM
+    /// (including Ok).
+    pub fn is_system_error(self) -> bool {
+        self.value < (Self::FIRST_UNRESERVED_EXIT_CODE)
+    }
+}
 
-    /// Indicates a message execution is forbidden for the caller.
-    SysErrForbidden = 8,
+impl std::fmt::Display for ExitCode {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.value)
+    }
+}
 
-    /// Indicates actor code performed a disallowed operation. Disallowed operations include:
-    /// - mutating state outside of a state acquisition block
-    /// - failing to invoke caller validation
-    /// - aborting with a reserved exit code (including success or a system error).
-    SysErrIllegalActor = 9,
+/// Enumerates exit codes which originate inside the VM.
+/// These values may not be used by actors when aborting.
+pub struct SystemExitCode {}
 
-    /// Indicates an invalid argument passed to a runtime method.
-    SysErrIllegalArgument = 10,
+impl SystemExitCode {
```

Comment:
Any reason not to attach these to `ExitCode`? I.e.

- `SYS_...`
- `USR_...`

The namespacing isn't great, but:

- Shorter names.
- Fewer imports.
- No constant-only structs (sometimes useful, but a bit strange).

Reply:
Ok, done.